x/group/server/invariants.go

```diff
@@ -0,0 +1,96 @@
+package server
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/regen-network/regen-ledger/orm"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/regen-network/regen-ledger/types"
+	"github.com/regen-network/regen-ledger/x/group"
+)
+
+func (s serverImpl) RegisterInvariants(ir sdk.InvariantRegistry) {
+	ir.RegisterRoute(group.ModuleName, "Tally-Votes", s.TallyVotesInvariant())
+}
+
+func (s serverImpl) AllInvariants() sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		return s.TallyVotesInvariant()(ctx)
+	}
+}
+
+func (s serverImpl) TallyVotesInvariant() sdk.Invariant {
+	return func(sdkCtx sdk.Context) (string, bool) {
+		var msg string
+		var broken bool
+		var curProposal []*group.Proposal
+		ctx := types.Context{Context: sdkCtx}
+		it1, err := s.proposalTable.PrefixScan(ctx, 1, math.MaxUint64)
+		if err != nil {
+			return "start value must be less than end value in iterator", false
+		}
+		curProposalRowID, err := orm.ReadAll(it1, &curProposal)
+		if err != nil {
+			return "cannot read all proposals in current block", false
+		}
+		_ = curProposalRowID
+		var prevProposal []*group.Proposal
+		if ctx.BlockHeight()-1 >= 0 {
+			sdkCtx = sdkCtx.WithBlockHeight(ctx.BlockHeight() - 1)
+		} else {
+			return "Not enough blocks to perform TallyVotesInvariant", false
+		}
+		it2, err := s.proposalTable.PrefixScan(sdkCtx, 1, math.MaxUint64)
+		if err != nil {
+			return "start value must be less than end value in iterator", false
+		}
+		prevProposalRowID, err := orm.ReadAll(it2, &prevProposal)
+		if err != nil {
+			return "cannot read all proposals from previous block", false
+		}
+		_ = prevProposalRowID
+		for i := 0; i < len(prevProposal) && i < len(curProposal); i++ {
+			if int32(prevProposal[i].Status) == 1 && int32(curProposal[i].Status) == 1 {
```

Comment:
why only verifying for this proposal status value?

Reply:
@blushi I believe here we are checking for only active proposals. Do we need to check all the proposals? I think that is not necessary

Reply:
@anilCSE yeah it makes sense since technically you cannot vote on a non active proposal (ie when `proposal.Status` different from `ProposalStatusSubmitted`) but I thought it would be more secure to double check everything.
In any case, if a proposal is closed at height `i` and it was still active at `i-1`, it's still relevant to compare them, since this means there has been a final tally for the proposal to get closed.